cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,370 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	//   )
+	DataKey string `json:"data-key"`
+	// TODO: make these values enums?
+	// DataState is the state of the ubuntu-data mountpoint, it can be:
+	// - "mounted" for mounted and available
+	// - "enc-not-found" for state where we found an ubuntu-data-enc
+	//   partition and tried to unlock it, but failed _before mounting it_.
+	//   Errors mounting are identified differently.
+	// TODO:UC20: "other" is a really bad term here replace with something better
+	// - "other" for some other error where we couldn't identify if there's an
+	//    ubuntu-data (or ubuntu-data-enc) partition at all
+	DataState string `json:"data-state"`
+	// SaveKey is which key was used to unlock ubuntu-save (if any). Same values
+	// as DataKey.
+	SaveKey string `json:"save-key"`
+	// SaveState is the state of the ubuntu-save mountpoint, it can be in any of
+	// the states documented for DataState, with the additional state of:
+	// - "not-needed" for state when we don't have ubuntu-save, but we don't need it (unencrypted only)
+	SaveState string `json:"save-state"`
+	// HostLocation is the location where the host's ubuntu-data is mounted
+	// and available. If this is the empty string, then the host's
+	// ubuntu-data is not available anywhere.
+	HostLocation string `json:"host-location"`
+}
+
+type context struct {
+	// the disk we have all our partitions on
+	disk disks.Disk
+	// options for finding partitions on the disk we have all our partitions on
+	diskOpts *disks.Options
+	// the device for ubuntu-data that is to be mounted - either an unencrypted
+	// volume, or a decrypted mapper volume if ubuntu-data is really encrypted
+	// on the physical disk
+	dataDevice string
+	// the device for ubuntu-save that is to be mounted
+	saveDevice string
+
+	// XXX: would be nice to get rid of tracking this here
+	isEncryptedDev bool
+
+	// state for tracking what happens
+	degradedState recoverDegradedState
+}
+
+// TODO: should this be a method on *context ?
+func verifyMountPointCtx(ctx *context, dir, name string) error {
+	matches, err := ctx.disk.MountPointIsFromDisk(dir, ctx.diskOpts)
+	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate boot: %s mountpoint is expected to be from disk %s but is not", name, ctx.disk.Dev())
+	}
+	return nil
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+type stateFunc func(ctx *context) (stateFunc, error)
+
+var (
+	// errStateDone is the error returned when the state machine is done
+	// executing without a critical error
+	errStateDone = fmt.Errorf("state machine done")
+)
+
+type stateMachine struct {
+	current stateFunc
+}
+
+func newStateMachine() *stateMachine {
+	m := &stateMachine{}
+	m.current = m.unlockDataRunKey
+	return m
+}
+
+func (m *stateMachine) execute(ctx *context) error {
```

Comment:
How about making this something like:
```
 func (m *stateMachine) execute(ctx *context) (finished bool, err error) {
        next, err := m.current(ctx)
	m.current = next
	return next == nil, err
}
```
to avoid the next to have a errStateDone  (which feels a bit strange to me).